internal/rpc/otel_interceptor.go

```diff
@@ -0,0 +1,156 @@
+		requestStartTime := time.Now()
+		isClient := request.Spec().IsClient
+		name := strings.TrimLeft(request.Spec().Procedure, "/")
+
+		attributes := getAttributes(ctx, request.Peer().Protocol)
+		spanKind := trace.SpanKindClient
+		requestSpan, responseSpan := semconv.MessageTypeSent, semconv.MessageTypeReceived
+		traceOpts := []trace.SpanStartOption{
+			trace.WithAttributes(attributes...),
+		}
+
+		if !isClient {
+			spanKind = trace.SpanKindServer
+			requestSpan, responseSpan = semconv.MessageTypeReceived, semconv.MessageTypeSent
+		}
+
+		traceOpts = append(traceOpts, trace.WithSpanKind(spanKind))
+		tracer := otel.GetTracerProvider().Tracer(request.Spec().Procedure)
+		ctx, span := tracer.Start(ctx, name, traceOpts...)
+		defer span.End()
+
+		var requestSize int
+		if request != nil {
+			if msg, ok := request.Any().(proto.Message); ok {
+				requestSize = proto.Size(msg)
+			}
+		}
+
+		span.AddEvent(otelFtlEventName,
+			trace.WithAttributes(
+				requestSpan,
+				semconv.MessageIDKey.Int(1),
+				semconv.MessageUncompressedSizeKey.Int(requestSize),
+			),
+		)
+		statusCode := attribute.Int64("rpc.grpc.status_code", 0)
+		response, err := next(ctx, request)
+		if err != nil {
+			statusCode = attribute.Int64(string(statusCode.Key), int64(connect.CodeOf(err)))
+		}
+		attributes = append(attributes, statusCode)
+		var responseSize int
+		if err == nil {
+			if msg, ok := response.Any().(proto.Message); ok {
+				responseSize = proto.Size(msg)
+			}
+		}
+		span.AddEvent(otelFtlEventName,
+			trace.WithAttributes(
+				responseSpan,
+				semconv.MessageIDKey.Int(1),
+				semconv.MessageUncompressedSizeKey.Int(responseSize),
+			),
+		)
+		span.SetAttributes(attributes...)
+
+		instrumentation, err := createInstruments(otel.GetMeterProvider().Meter("ftl.rpc.unary"))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create instruments: %w", err)
+		}
+		instrumentation.duration.Record(ctx, time.Since(requestStartTime).Milliseconds(), metric.WithAttributes(attributes...))
+		instrumentation.requestSize.Record(ctx, int64(requestSize), metric.WithAttributes(attributes...))
+		instrumentation.responseSize.Record(ctx, int64(responseSize), metric.WithAttributes(attributes...))
+		return response, err
+	}
+}
+
+func (i *otelInterceptor) WrapStreamingClient(next connect.StreamingClientFunc) connect.StreamingClientFunc {
+	return func(ctx context.Context, spec connect.Spec) connect.StreamingClientConn {
+		return next(ctx, spec)
+	}
+}
+
+func (i *otelInterceptor) WrapStreamingHandler(next connect.StreamingHandlerFunc) connect.StreamingHandlerFunc {
+	return func(ctx context.Context, conn connect.StreamingHandlerConn) error {
+		return next(ctx, conn)
+	}
+}
+
+type instruments struct {
+	duration     metric.Int64Histogram
+	requestSize  metric.Int64Histogram
+	responseSize metric.Int64Histogram
+}
+
+func createInstruments(meter metric.Meter) (instruments, error) {
```

Comment:
It would be nice if we didn't lose _all_ the instruments whenever one of their instantiations fails. What do you think about refactoring the code here and above to look something like below? This would also have the helpful maintainability effect of parceling the computation of each metric value into its own function, so it's easier to figure out what logic is used for what. `WrapUnary` is somewhat long at the moment, and it's easy to imagine how it could grow longer.

```
func (i *otelInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
	return func(ctx context.Context, request connect.AnyRequest) (connect.AnyResponse, error) {
		...
		recordDuration(ctx, attributes, requestStartTime)
		recordRequestSize(...)
		recordResponseSize(...)
		return response, err
	}
}

func recordDuration(ctx context.Context, attributes []attribute.KeyValue, requestStartTime time.Time) {
	recordHist(ctx, attributes, histDetails{
		name: "duration",
		unit: "ms",
		desc: "...",
		val: time.Since(requestStartTime).Milliseconds(),
	})
}

type histDetails struct {
	name string
	unit string
	desc string
	val  int64
}

func recordHist(ctx context.Context, attributes []attribute.KeyValue, details histDetails) {
	logger := log.FromContext(ctx) // just log the error instead of exiting the function early
	hist, err := meter.Int64Histogram(details.name, metric.WithUnit(details.unit), metric.WithDescription(details.desc))
	if err != nil {
		logger.Errorf(...)
	}
	hist.Record(ctx, details.val, metric.WithAttributes(attributes...))
}
```

Reply:
Good idea!

Reply:
We ended up deciding to panic on metric instantiation error after all to make sure we don't miss any errors in development, but the rest of this still holds :)